symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit.hs

```diff
@@ -33,46 +33,115 @@ module ZkFold.Symbolic.Compiler.ArithmeticCircuit (
+import           Text.Pretty.Simple                                      (pPrint)
 
 import           ZkFold.Base.Algebra.Basic.Class
-import           ZkFold.Base.Algebra.Polynomials.Multivariate        (evalMonomial, evalPolynomial)
-import           ZkFold.Base.Data.HFunctor                           (hmap)
-import           ZkFold.Base.Data.Product                            (fstP, sndP)
-import           ZkFold.Prelude                                      (length)
-import           ZkFold.Symbolic.Class                               (fromCircuit2F)
-import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance ()
-import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal (Arithmetic, ArithmeticCircuit (..), Constraint,
-                                                                      SysVar (..), Var (..), acInput, eval, eval1, exec,
-                                                                      exec1, hlmap, witnessGenerator)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate            (evalMonomial, evalPolynomial)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..))
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..))
+import           ZkFold.Base.Data.HFunctor                               (hmap)
+import           ZkFold.Base.Data.Product                                (fstP, sndP)
+import           ZkFold.Prelude                                          (length)
+import           ZkFold.Symbolic.Class                                   (fromCircuit2F)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal     (Arithmetic, ArithmeticCircuit (..),
+                                                                          Constraint, SysVar (..), Var (..),
+                                                                          WitVar (..), acInput, eval, eval1, exec,
+                                                                          exec1, hlmap, witnessGenerator)
 import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Map
-import           ZkFold.Symbolic.Data.Combinators                    (expansion)
-import           ZkFold.Symbolic.MonadCircuit                        (MonadCircuit (..))
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+import           ZkFold.Symbolic.Data.Combinators                        (expansion)
+import           ZkFold.Symbolic.MonadCircuit                            (MonadCircuit (..))
 
 --------------------------------- High-level functions --------------------------------
 
 -- | Optimizes the constraint system.
 --
--- TODO: Implement nontrivial optimizations.
-optimize :: ArithmeticCircuit a p i o -> ArithmeticCircuit a p i o
-optimize = id
+optimize1 :: forall a p i o. ( Arithmetic a, Ord (Rep i))
```

Comment:
Please add the description of what it does.

Reply:
Also I don't think that the State monad is needed here, it can be expressed as just `SysVar i -> a -> AC a p i o -> AC a p i o`